### PR TITLE
Feat: deposits for Eth Network

### PR DESCRIPTION
**Pull request by ayush-tanx**

### CODE REVIEW

**Comment by austinjfi** [comment]
Hey @ayush-tanx, please use [pip-tools](https://pip-tools.readthedocs.io/en/latest/) to create the requirements.txt and dev-requirements.txt. The command is given at the top of the requirements.txt file.
And make sure that requirements for development and production is kept separate.

File: src/brineconnector/client.py [hunk 1]
```diff
@@ -1,17 +1,49 @@
...
+    CoinStatPayload
+)
+from web3 import Web3, Account
+from .constants import Config
+from decimal import Decimal
+from dotenv import load_dotenv
+import os
+load_dotenv()
 
+rpc_provider = os.environ['RPC_PROVIDER']
```
Comment (austinjfi) [inline src/brineconnector/client.py:35]: Please don't add env variables in client (users wont be able to use this, please think from the perspective that youre installing a pip package), dotenv is only supposed to be show as an example. Check how its done in the nodejs connector, it should only be passed in as parameters.

File: requirements.txt [hunk 2]
```diff
@@ -1,130 +1,66 @@
...
+parsimonious==0.8.1
+pkgutil_resolve_name==1.3.10
+platformdirs==4.2.0
+pluggy==1.4.0
+protobuf==3.20.3
 pycryptodome==3.17
-    # via
-    #   eth-hash
-    #   eth-keyfile
+pylint==3.0.3
```
Comment (austinjfi) [inline requirements.txt:45]: Is pylint required?

File: requirements.txt [hunk 3]
```diff
@@ -1,130 +1,66 @@
...
 pyrsistent==0.19.3
-    # via jsonschema
+pytest==8.0.0
+python-dotenv==1.0.1
```
Comment (austinjfi) [inline requirements.txt:48]: Can you generate the requirement files properly.
There are dependencies here which aren't required to be installed by users.
Start with dependencies and dev-dependencies in pyproject.toml and install them in separate virtual environments. And then use pip tools to create the requirement files.

File: src/brineconnector/bin/python_signature.py
Comment (austinjfi) [inline src/brineconnector/bin/python_signature.py:1]: Can you check if functions in this file is being used anywhere?
If not please remove it.

File: src/brineconnector/helpers.py
Comment (austinjfi) [inline src/brineconnector/helpers.py:1]: Adding to [this](https://github.com/tanx-libs/tanx-connector-python/pull/1/files#r1505392121). Please check if this file is being used anywhere as well.

Comment (austinjfi) [inline src/brineconnector/helpers.py:1]: Adding to [this](https://github.com/tanx-libs/tanx-connector-python/pull/1/files#r1505392121). Please check if this file is being used anywhere as well.

File: src/brineconnector/utils.py [hunk 4]
```diff
@@ -22,7 +25,7 @@ def sign_msg_hash(nonce: CreateOrderNoncePayload, private_key: str, option: Lite
 
 
 def create_user_signature(private_key: str, option: Literal['mainnet', 'testnet'] = 'mainnet') -> str:
```
Comment (austinjfi) [inline src/brineconnector/utils.py:27]: Please remove this function, to prevent stark key mismatch.

File: tests/test_client.py [hunk 5]
```diff
@@ -385,3 +385,29 @@ def test_list_trades():
...
+                        'message':'Essential parameters are missing',
+                        'payload': ''})
+    with pytest.raises(requests.exceptions.RequestException):
+        res = client.crypto_deposit_start('100000','0x27..','0x27..','0x67..','930',65707,)
+    # Handle exception
+        
+        data = res.response.json()
+        assert 'status' in data
+        assert data['status'] == 'error'
+        assert 'Essential parameters' in data['message']
```
Comment (austinjfi) [inline tests/test_client.py:413]: Can you research and add anymore edgecases?
Such as gas limit exceeded, not enough balance, etc

**Comment by austinjfi** [comment]
Can you fix the github actions as well, seems to be some dependency issue

**Comment by austinjfi** [comment]
Can you also add documentation in the README on how to use deposits?

**Review by austinjfi — CHANGES_REQUESTED** [review]
Please have a look at the comments

File: src/brineconnector/utils.py (hunk 4 above)
Comment (ayush-tanx) [reply src/brineconnector/utils.py:27]: I'm keeping this method for current use and local testing, was thinking to remove this once I start to work with nodejs one completely

File: example/rest.py [hunk 6]
```diff
@@ -55,4 +74,61 @@ def main():
...
+            print(exc.response.json())
+
+        user_signature = create_user_signature(PRIVATE_KEY, 'testnet')
+        key_pair = get_stark_key_pair_from_signature(user_signature)
+        stark_public_key = key_pair['stark_public_key']
+        provider = Web3(Web3.HTTPProvider(rpc_provider))
+        signer = Account.from_key(PRIVATE_KEY)
+
+        # # deposit with eth private key
+        # deposit_res_with_private_key = client.deposit_from_ethereum_network(
```
Comment (austinjfi) [inline example/rest.py:100]: Running this function gives this error (moved outside try except block to get traceback)
```
<class 'eth_account.signers.local.LocalAccount'>
Traceback (most recent call last):
  File "example/rest.py", line 134, in <module>
    ethereumDeposit()
  File "example/rest.py", line 125, in ethereumDeposit
    deposit_res_with_private_key = client.deposit_from_ethereum_network(
  File "/home/austin/Documents/brine/tanx-connector-python/src/brineconnector/client.py", line 345, in deposit_from_ethereum_network
    return self.deposit_from_ethereum_network_with_starkKey(signer, provider, f'0x{stark_public_key}', amount, currency)
  File "/home/austin/Documents/brine/tanx-connector-python/src/brineconnector/client.py", line 273, in deposit_from_ethereum_network_with_starkKey
    parsed_amount = w3.toWei(amount, 'ether')
AttributeError: 'Web3' object has no attribute 'toWei'
```

File: src/brineconnector/client.py [hunk 7]
```diff
@@ -164,3 +185,157 @@ def list_trades(self, limit: Optional[int] = None, page: Optional[int] = None, m
...
+            deposit_response['hash'].hex(),
+            transaction['nonce'],
+            vault['payload']['id']
+        )
+
+        res['payload'] = {'transaction_hash': deposit_response['hash'].hex()}
+
+        return res
+
+    def deposit_from_ethereum_network(self, rpc_url: str, eth_private_key: str, network: Literal['mainnet', 'testnet'], currency: str, amount: float):
```
Comment (austinjfi) [inline src/brineconnector/client.py:333]: remove this function

File: src/brineconnector/client.py [hunk 8]
```diff
@@ -164,3 +185,147 @@ def list_trades(self, limit: Optional[int] = None, page: Optional[int] = None, m
...
+    def approve_unlimited_allowance_ethereum_network(self, coin: str, signer, w3: Web3):
+        coin_stats = self.get_coin_stats()
+        current_coin = filter_ethereum_coin(coin_stats['payload'], coin)
+        token_contract = current_coin['token_contract']
+        stark_contract = Config.STARK_CONTRACT[self.option]
+        res = approve_unlimited_allowance_util(contract_address=stark_contract, token_contract=token_contract, signer=signer, w3=w3)
+        return res
+
+
+    def deposit_from_ethereum_network_with_starkKey(self, signer: Account, provider: Web3, stark_public_key: str, amount: float, currency: str):
```
Comment (austinjfi) [inline src/brineconnector/client.py:245]: Change starkKey to stark_key

**Review by austinjfi — APPROVED** [review]

File: README.md [hunk 9]
```diff
@@ -316,3 +316,44 @@ except requests.exceptions.HTTPError as exc:
...
+#### Ethereum Deposit
+
+There are two ways to make a deposit on the Ethereum network:
+
+<!-- 1. Using ETH Private Key and RPC URL: This approach utilizes your ETH private key and an rpcUrl (e.g., from Infura or Alchemy).
+2. Custom Provider and Signer: This method involves creating your provider and signer using web3.py. You'll also need the stark_public_key. -->
+
+#### Using Custom Provider and Signer:
+
+This method involves using a custom provider and signer, which can be created using the web3.py library. The `stark_public_key` mentioned in the code should be obtained using the steps described in the [Create L2 Key Pair](#create-l2-key-pair) section of the nodejs connector(https://github.com/tanx-libs/tanx-connector-nodejs#create-l2-key-pair). Here's the code snippet for this method:
```
Comment (austinjfi) [inline README.md:330]: Create L2 Key Pair points to the same repo. It should point to the nodejs connector

**Pull request merged**